docs/design/upgrade.md

```diff
@@ -0,0 +1,203 @@
+
+In all those cases, the input of the upgrade function is a project (which includes configuration, packages, root tree, build files, ...) and the output will be a list of changes. These changes might be applied or not, but the upgrade function should not have any side effects.
+
+### Code
+An AST-to-AST transformation step will be necessary to perform complex refactoring. The internal `upgrade` library can be used to build a transformation pipeline that will be applied to every source files in the Angular app.
+
+The pipeline is a stream of `ts.SourceFile`, and should receive a list of changes to apply to those sources.
+
+A transform function is a change or an array of changes. See [Change API](#change-api).
+
+There are helpers exported to create an RxJs stream of nodes and provide changes easily:
+
+```typescript
+import {Observable} from 'rxjs/Rx';
+import {runRefactor, visitNodes} from 'upgrade/api';
+import {renameNode} from 'upgrade/api/changes';
+import {anyOf, isClass, isFunction} from 'upgrade/api/matchers';
+import * as ts from 'typescript';
+
+function renameFunctions(source: ts.SourceFile): Observable<Change> {
+  // `visitNodes` return all the nodes in the tree.
+  return source.let(visitNodes)
+    .filter(anyOf(isClass, isFunction))
+    .flatMap((node: ts.Node): Change[] => {
+      // Rename a node.
+      return [renameNode(node, (<ts.Identifier>node).text + 'Upgraded')];
+    });
+}
+
+export function upgrade(project): Observable<Change> {
+  return runRefactor(renameFunctions);
+}
+```
+
+This would apply to all classes and all functions in every source files, which might be too much. It is only shown as an example.
+
+### <a name="change-api"></a>Change API
+A change is an operation that happens on string. It is an object that implements the following interface:
+
+```typescript
+interface Change {
+  // Apply the change.
+  apply(): Thenable<boolean>;
+
+  // The file this change should be applied to. Some changes might not apply to
+  // a file (maybe the config).
+  readonly path: string | null;
+
+  // The order this change should be applied. Normally the position inside the file.
+  // Changes are applied from the bottom of a file to the top.
+  readonly order: number | null;
+
+  // The description of this change. This will be outputted in a dry or verbose run.
+  readonly description: string;
+}
+```
+
+Some available example of a change implementation:
+
+```typescript
+// Will add text to the source code.
+class AppendChange implements Change {
+  constructor(file: string, pos: number, toAdd: string);
+}
+// Will remove text from the source code.
+class RemoveChange implements Change {
+  constructor(file: string, pos: number, toRemove: string);
+}
+// Will replace text from the source code.
+class ReplaceChange implements Change {
+  constructor(file: string, pos: number, oldText: string, newText: string);
+}
+// Will output a message for the user to fulfill.
+class MessageChange implements Change {
```

Comment:
How do we verify/recover from incorrect user input? According to atomicity, if compilation fails it's back to the drawing board. Perhaps we should have a better way.